src/mem_map.c

```diff
@@ -304,8 +306,30 @@ static int mem_region_map_file(struct mem_region *region)
         return -1;
     }
 
-    data = mmap(NULL, region->length, PROT_READ,
-            MAP_SHARED, region->fd, region->offset);
+    if (fstat(region->fd, &stat_buf) < 0) {
+        int err = errno;
+        fprintf(stderr, "failed to stat file %s: %s\n", region->path, strerror(err));
+        return -1;
+    }
+
+    if (region->offset > stat_buf.st_size) {
+        fprintf(stderr, "mmap offset too large (%d > %d)\n", region->offset, stat_buf.st_size);
+        return -1;
+    }
+
+    // Accessing beyond the length of the file, even though we can map a
+    // region larger than the size of the file, will cause a SIGBUS, so
+    // truncate the length of the map to fit within the file.
+    if (region->length > stat_buf.st_size - region->offset) {
+        fprintf(stderr, "mmap length > size - offset (%d > %d), truncating "
```

Comment:
Should we always print this message? It seems to be a problem within a program, not ours.